pkg/api/types.go

```diff
@@ -1546,6 +1564,119 @@ func (p *Properties) GetCustomCloudIdentitySystem() string {
 	return AzureADIdentitySystem
 }
 
+// SetCustomCloudProfileEnvironment retrives the endpoints from Azure Stack metadata endpoint and set the values for azure.Environment
+func (cs *ContainerService) SetCustomCloudProfileEnvironment() {
+	p := cs.Properties
+	if p.IsAzureStackCloud() {
+		env := p.CustomCloudProfile.Environment
+		if env.ServiceManagementEndpoint == "" || env.ActiveDirectoryEndpoint == "" || env.GraphEndpoint == "" || env.ResourceManagerVMDNSSuffix == "" {
+			metadataURL := fmt.Sprintf("%s/metadata/endpoints?api-version=1.0", strings.TrimSuffix(env.ResourceManagerEndpoint, "/"))
+
+			// Retrieve the metadata
+			httpClient := &http.Client{
+				Timeout: 15 * time.Second,
+			}
+			endpointsresp, err := httpClient.Get(metadataURL)
+			if err != nil || endpointsresp.StatusCode != 200 {
+				log.Fatalf("%s . apimodel invalid: failed to retrieve Azure Stack endpoints from %s", err, metadataURL)
+			}
+
+			body, err := ioutil.ReadAll(endpointsresp.Body)
+			if err != nil {
+				log.Fatalf("%s . apimodel invalid: failed to read the response from %s", err, metadataURL)
```

Comment:
Instead of logging errors and exiting here, make the function return type an error and return the error. The only place that should exit and log the error is the original caller, main.go. 

Same comment for all the other `log.Fatalf` calls in the code.